Re-enable --enable-hardening security features

When we added the `-O0` flag to disable optimization, that disabled `FORTIFY_SOURCE` security features. You can see the warnings saying it requires optimization to be enabled in the build output. We have to re-enable optimization (and make sure `--enable-hardening` is on) before we release 1.0.

```
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/date_time/wrapping_int.hpp:151:14: error: conversion to 'boost::date_time::wrapping_int2<short int, 1, 12>::int_type {aka short int}' from 'int' may alter its value [-Werror=conversion]
       value_ -= (wrap_max - wrap_min + 1);
       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/firexware/Data-1/Documents/FireXware/la/zcash/depends/x86_64-unknown-linux-gnu/include/boost/date_time/wrapping_int.hpp:156:14: error: conversion to 'boost::date_time::wrapping_int2<short int, 1, 12>::int_type {aka short int}' from 'int' may alter its value [-Werror=conversion]
       value_ += (wrap_max - wrap_min + 1);
       ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
configure:26641: $? = 1
configure: failed program was:
```

--- a/configure.ac
+++ b/configure.ac
@@ -413,7 +413,7 @@ fi
 
 if test x$use_hardening != xno; then
   AX_CHECK_COMPILE_FLAG([-Wformat],[HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -Wformat"],[AC_MSG_ERROR(Cannot enable -Wformat)])
-  AX_CHECK_COMPILE_FLAG([-Wformat-security],[HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -Wformat-security"],[AC_MSG_ERROR(Cannot enable -Wformat-security)])
+  AX_CHECK_COMPILE_FLAG([-Wformat-security],[HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -Wformat-security"],[AC_MSG_ERROR(Cannot enable -Wformat-security)],[-Wformat])
   AX_CHECK_COMPILE_FLAG([-Wstack-protector],[HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -Wstack-protector"],[AC_MSG_ERROR(Cannot enable -Wstack-protector)])
   AX_CHECK_COMPILE_FLAG([-fstack-protector-all],[HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -fstack-protector-all"],[AC_MSG_ERROR(Cannot enable -fstack-protector-all)])
 

--- a/configure.ac
+++ b/configure.ac
@@ -412,6 +412,8 @@ if test x$use_glibc_compat != xno; then
 fi
 
 if test x$use_hardening != xno; then
+  AX_CHECK_COMPILE_FLAG([-Wformat],[HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -Wformat"],[AC_MSG_ERROR(Cannot enable -Wformat)])
+  AX_CHECK_COMPILE_FLAG([-Wformat-security],[HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -Wformat-security"],[AC_MSG_ERROR(Cannot enable -Wformat-security)])
   AX_CHECK_COMPILE_FLAG([-Wstack-protector],[HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -Wstack-protector"],[AC_MSG_ERROR(Cannot enable -Wstack-protector)])
   AX_CHECK_COMPILE_FLAG([-fstack-protector-all],[HARDENED_CXXFLAGS="$HARDENED_CXXFLAGS -fstack-protector-all"],[AC_MSG_ERROR(Cannot enable -fstack-protector-all)])
 

